File: 402-remove-k-digits/402-remove-k-digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack=[]
        for i in num:
            while k>0 and stack and int(i)<int(stack[-1]):
                stack.pop()
                k-=1
            stack.append(i)
        #incase the number was in increasing order, 
        #we just take the left most digits
        res=''.join(stack[:len(stack)-k]) 
        return str(int(res)) if res else '0'

File: 402-remove-k-digits/402-remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        Stack<Character> stack = new Stack<Character>();
        int i=0;
        while(i<num.length()){
            while (k>0 && !stack.isEmpty() && num.charAt(i)<stack.peek()){
                stack.pop();
                k--;
            }
            if (!stack.isEmpty() || num.charAt(i)!='0'){
            stack.push(num.charAt(i));
            }
            i++;
        }
        while(k>0 && !stack.isEmpty()){
            stack.pop();
            k--;
        }
        StringBuilder result=new StringBuilder();
        while (!stack.isEmpty()){
            result.append(stack.pop());
        }
        result.reverse();
        String temp=result.toString();
        if (temp==""){
            return "0";
        }
        return temp;
    }
}